File: server/src/Signal/Signal.java
package Signal;

import org.json.JSONException;

import java.util.ArrayList;
import java.util.Vector;

public class Signal<T> {
    private ArrayList<Event> events;
    public Signal(){
        events = new ArrayList();
    }

    public Connection connect(Event event){
        Connection conn = new Connection(() -> {
            events.remove(event);
        });
        events.add(event);

        return conn;
    }

    public void fire(T arg) throws Exception {
        for (Event event: events) {
            event.run(arg);
        }
    }

    public void fireAsync(T arg) throws InterruptedException {
        Vector<Thread> threads = new Vector<>();
        for (Event event: events) {
            Thread th = new Thread(()->{
                try {
                    event.run(arg);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            });
            threads.add(th);
            th.start();
        }

        for(Thread th : threads)
            th.join();
    }
}

File: client/js/Signal/Signal.js
class Connection {
    signal;
    id;

    constructor(signal, id) {
        this.signal = signal;
        this.id = id;
    }

    disconnect(){
        delete this.signal.connections[this.id];
        this.signal.connections[this.id] = null;
    }
}
export class Signal {
    connections;
    curr_id = 0;

    constructor() {
        this.connections = {};
    }

    connect(fn){
        ++this.curr_id;
        this.connections[this.curr_id] = fn;

        return new Connection(this, this.curr_id);
    }

    fire(...args){
        for(let key in this.connections)
            this.connections[key](args);
    }

    clear(){
        this.connections = {};
    }
}